File: subdiv.m
gsx = 256;
gsy = 256;
gsz = 256;

lsx = 128;
lsy = 128;
lsz = 128;
FILE = fopen('Data/foot.raw');
A = fread(FILE,'uint8=>uint8');
B = reshape(A,gsx,gsy,gsz);
S = strcat('parallel',{' '},':::',{' '});
%sd = gs/ls;
 %C = cell((gsx/lsx)^3,1); %assuming each subdomain ios of equal size
 count =1;
%B=F;
for z = 1:lsz:(gsz-lsz+1)
    for y =1:lsy:(gsy-lsy+1)
        for x=1:lsx:(gsx-lsx+1)
            if (x==(gsx-lsx+1)) 
                cx = 0;
            else cx = 1;
            end
            if (y==(gsy-lsy+1)) 
                cy = 0;
            else cy =1;
            end
            if (z==(gsz-lsz+1)) cz = 0;
            else cz =1;
            end
            %fopen'w'
            filestr = (strcat('Data/fo_',num2str(count),'_',num2str(lsy+cy),'_',num2str(lsx+cx),'_',num2str(lsz+cz),'.raw'));
            FILE = fopen(filestr,'w');
            S = strcat(S,'"./serialct',{' '},filestr,{' '}, num2str(lsy+cy),{' '},num2str(lsx+cx),{' '},num2str(lsz+cz),{' '},'f',num2str(count),{' '},num2str(count),'"', {' '}); 
            
            fwrite(FILE,B(y:y+ls-1+cy,x:x+ls-1+cx,z:z+ls-1+cz),'uint8');
            %C{count} = B(y:y+ls-1+cy,x:x+ls-1+cx,z:z+ls-1+cz);
            count = count+1
        end
    end
end
scriptfile = fopen('Data/tree.sh','w');
fprintf(scriptfile,'%s',S{1});

File: subdiv.py
import numpy as np
import os
import sys
import array

filename = str(sys.argv[1])
gsx = int(sys.argv[2])
gsy = int(sys.argv[3])
gsz = int(sys.argv[4])

lsx = int(sys.argv[5])
lsy = int(sys.argv[6])
lsz = int(sys.argv[7])

input_file = open(filename,'r')
parent_folder = os.path.dirname(input_file.name)
A = np.fromfile(input_file, dtype='uint8')
B = np.reshape(A, (gsx, gsy, gsz))

S = 'parallel ::: '

count = 1

for z in range(0,(gsz-lsz+1),lsz):
    for y in range(0,(gsy-lsy+1),lsy):
        for x in range(0,(gsx-lsx+1),lsx):
            if (x==(gsx-lsx)):
                cx = 0
            else:
                cx = 1
            if (y==(gsy-lsy)):
                cy = 0
            else:
                cy = 1
            if (z==(gsz-lsz)):
                cz = 0
            else:
                cz = 1
            stc = str(count)
            stx = str(lsx+cx)
            sty = str(lsy+cy)
            stz = str(lsz+cz)
            filestr = parent_folder + os.sep + 'fo_' + stc + '_' + stx + '_' + sty + '_' + stz + '.raw'
            S = S + '"./serialct '+ filestr + ' ' + stx + ' ' + sty + ' ' + stz + ' f' + stc + ' ' + stc + '"' + ' '
            C = B[x: x+lsx+cx, y: y+lsy+cy, z: z+lsz+cz]
            C.tofile(filestr)
            count += 1
with open(parent_folder + os.sep +'tree.sh', "w") as tree:
    tree.write(S)
with open(parent_folder + os.sep +'params.txt', "w") as params:
    params.write(str(gsx/lsx) + ' ' + str(gsy/lsy) + ' ' + str(gsz/lsz) + ' '+ str(lsx) + ' ' + str(lsy) + ' ' + str(lsz))
input_file.close()
